src/main/scala/is/hail/annotations/MemoryBlock.scala

```diff
@@ -571,65 +588,61 @@ class RegionValueBuilder(region: MemoryBuffer) {
     val i = indexstk.top
     typestk.top match {
       case t: TStruct =>
-        region.setBit(offsetstk.top, i)
+        t.setFieldMissing(region, offsetstk.top, i)
       case t: TArray =>
-        region.setBit(offsetstk.top + 4, i)
+        t.setElementMissing(region, offsetstk.top, i)
     }
 
     advance()
   }
 
   def addBoolean(b: Boolean) {
-    current() match {
-      case (TBoolean, off) =>
-        region.storeByte(off, b.toByte)
-        advance()
-    }
+    assert(currentType() == TBoolean)
+    val off = currentOffset()
+    region.storeByte(off, b.toByte)
+    advance()
   }
 
   def addInt(i: Int) {
-    current() match {
-      case (TInt32, off) =>
-        region.storeInt(off, i)
-        advance()
-    }
+    assert(currentType() == TInt32)
+    val off = currentOffset()
+    region.storeInt(off, i)
+    advance()
   }
 
   def addLong(l: Long) {
-    current() match {
-      case (TInt64, off) =>
-        region.storeLong(off, l)
-        advance()
-    }
+    assert(currentType() == TInt64)
+    val off = currentOffset()
+    region.storeLong(off, l)
+    advance()
   }
 
   def addFloat(f: Float) {
-    current() match {
-      case (TFloat32, off) =>
-        region.storeFloat(off, f)
-        advance()
-    }
+    assert(currentType() == TFloat32)
+    val off = currentOffset()
+    region.storeFloat(off, f)
+    advance()
   }
 
   def addDouble(d: Double) {
-    current() match {
-      case (TFloat64, off) =>
-        region.storeDouble(off, d)
-        advance()
-    }
+    assert(currentType() == TFloat64)
+    val off = currentOffset()
+    region.storeDouble(off, d)
+    advance()
   }
 
   def addBinary(bytes: Array[Byte]) {
-    current() match {
-      case (TBinary, off) =>
-        region.align(4)
-        val boff = region.offset
-        region.storeAddress(off, boff)
-
-        region.appendInt(bytes.length)
-        region.appendBytes(bytes)
-        advance()
-    }
+    assert(currentType() == TBinary)
+    val off = currentOffset()
+
+    region.align(TBinary.contentAlignment)
+    val boff = region.offset
+    region.appendInt(bytes.length)
+    region.appendBytes(bytes)
+
+    region.storeAddress(off, boff)
```

Comment:
This still has the problem where storing just a TBinary in a region (with no container) will overwrite the first N bits with the offset, right?